keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,401 @@
+
+A CSI Node Plugin will interact with csi-proxy.exe using a named pipe: `\\.\pipe\csi-proxy` (exposed by csi-proxy.exe). The pipe will need to be mounted into the Node Plugin container from the host using the pod's volume mount specifications. Note that domain sockets cannot be used in this scenario since Windows blocks a containerized process from interacting with a host process that is listening on a domain socket. 
+
+A GRPC based interface will be used by CSI Node Plugins to invoke privileged operations through csi-proxy.exe. The GRPC API will be versioned and any release of csi-proxy.exe binary will maintain backward compatibility across all prior versions of the API. This avoids having to run multiple versions of the csi-proxy.exe binary on the same Windows host if multiple CSI Node Plugins have been configured and the version of the csi-proxy API required by the plugins are wide apart.
+
+The following are the main RPC calls that will comprise a v1 version of the csi-proxy API:
+
+```
+service CSIProxyService {
+
+    // Version returns the earliest and latest version of the CSI Proxy API supported by the service.
+    rpc Version(VersionRequest) returns (VersionResponse) {}
+
+    // Mkdir creates a directory at the requested absolute path in the host. Relative path is not supported.
+    rpc Mkdir(MkdirRequest) returns (MkdirResponse) {}
+
+    // Rmdir removes the directory at the requested absolute path in the host. Relative path is not supported.
+    // This may be used for unlinking a symlink created through LinkVolume
+    rpc Rmdir(RmdirRequest) returns (RmdirResponse) {}
+
+    // StageDisk partitions a disk device and formats the partition with NTFS (if the disk has not been partitioned or formatted)
+    // The resulting volume is mounted at the requested global staging path.
+    rpc StageDisk(StageDiskRequest) returns (StageDiskResponse) {}
+
+    // StageSMBShare mounts a remote share over SMB on the host at the requested global staging path.
+    rpc StageSMBShare(StageSMBShareRequest) returns (StageSMBShareResponse) {}
+
+    // LinkVolume invokes mklink on the global staging path of a volume linking it to a path within a container
+    rpc LinkVolume(LinkVolumeRequest) returns (LinkVolumeResponse) {}
+
+    // GetDiskNumberWithLocation finds a disk device with requested Location <Bus, Target, LUN ID>
+    rpc GetDiskNumberWithLocation(GetDiskNumberWithLocationRequest) returns (GetDiskNumberWithLocationResponse) {}
```

Comment:
What if the Location information isn't available anywhere? The AWS EBS driver would like to run a `ebsnvme-id.exe` on the host to map the cloud's reported volume-ID to the disk number: https://github.com/kubernetes/kubernetes/pull/79552/files#diff-238adf22aa109b4ad41399fe48280ec8R214. GCE PD may be similar, it runs some powershell cmdlets on the host: https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/gcepd/attacher.go#L147

Adding a proxy call to `Get-Disk` to the API will be a suitable workaround for EBS (ebsnvme-id.exe is not strictly necessary, at the moment `Get-Disk` is guaranteed to return the same information), but I'm not sure about GCE PD.

Reply:
@jingxu97 to determine if this api works for GCE PD or we need a special service to map PD devices.

Reply:
As part of the prototype to make sure everything does work end-2-end from a OS plumbing perspective, I ported the GCE PD CSI driver to work on Windows node. As part of that, I got the SCSI Page 83 queries and mapping working (required for the mapping of GCE PD and potentially by other vendors too) by creating a generic version of the `Get-GcePdName` cmdlet that directly issues `IOCTL_STORAGE_QUERY_PROPERTY`. The generic wrapper for that is exposed through the `GetDiskNumberWithSCSIID` API mentioned in the KEP. 

For the `GetDiskNumberWithLocation` API, I was indeed planning to use the information returned through `Get-Disk` @wongma7.

Reply:
I smiled when I saw `GetDiskNumberWithSCSIID`, I knew exactly why it was needed...

The GCE Persistent Disk driver (in-tree at least, haven't looked at the CSI version but assuming it's the same...) needs to map the PD "friendly name" to a Windows device id. We're starting to investigate if we can provide this information natively through the `Get-Disk` cmdlet (https://github.com/kubernetes/kubernetes/issues/74674). Even if we can change GCE to support this though, it looks like the PD driver would still end up using `GetDiskNumberWithSCSIID` rather than `GetDiskNumberWithLocation` since the latter will use <Bus, Target, LUN ID> as its key. Is that right?

Is the PD driver the only storage driver that uses such a "friendly name" for identifying the volume the pod wants to use? If not then perhaps an API method that accepts that as the argument would be useful; if so then `GetDiskNumberWithSCSIID` seems probably suitable.